flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/catalog/AbstractJDBCCatalog.java

```diff
@@ -0,0 +1,277 @@
+import org.apache.flink.table.catalog.exceptions.DatabaseNotExistException;
+import org.apache.flink.table.catalog.exceptions.FunctionAlreadyExistException;
+import org.apache.flink.table.catalog.exceptions.FunctionNotExistException;
+import org.apache.flink.table.catalog.exceptions.PartitionAlreadyExistsException;
+import org.apache.flink.table.catalog.exceptions.PartitionNotExistException;
+import org.apache.flink.table.catalog.exceptions.PartitionSpecInvalidException;
+import org.apache.flink.table.catalog.exceptions.TableAlreadyExistException;
+import org.apache.flink.table.catalog.exceptions.TableNotExistException;
+import org.apache.flink.table.catalog.exceptions.TableNotPartitionedException;
+import org.apache.flink.table.catalog.exceptions.TablePartitionedException;
+import org.apache.flink.table.catalog.stats.CatalogColumnStatistics;
+import org.apache.flink.table.catalog.stats.CatalogTableStatistics;
+import org.apache.flink.table.expressions.Expression;
+import org.apache.flink.table.factories.TableFactory;
+import org.apache.flink.util.StringUtils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.SQLException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/**
+ * Abstract catalog for any JDBC catalogs.
+ */
+public abstract class AbstractJDBCCatalog extends AbstractCatalog {
+
+	private static final Logger LOG = LoggerFactory.getLogger(AbstractJDBCCatalog.class);
+
+	protected final String username;
+	protected final String pwd;
+	protected final String baseUrl;
+	protected final String defaultUrl;
+
+	public AbstractJDBCCatalog(String catalogName, String defaultDatabase, String username, String pwd, String baseUrl) {
+		super(catalogName, defaultDatabase);
+
+		checkArgument(!StringUtils.isNullOrWhitespaceOnly(username));
+		checkArgument(!StringUtils.isNullOrWhitespaceOnly(pwd));
+		checkArgument(!StringUtils.isNullOrWhitespaceOnly(baseUrl));
+
+		JDBCCatalogUtils.validateJDBCUrl(baseUrl);
+
+		this.username = username;
+		this.pwd = pwd;
+		this.baseUrl = baseUrl.endsWith("/") ? baseUrl : baseUrl + "/";
+		this.defaultUrl = baseUrl + defaultDatabase;
+	}
+
+	@Override
+	public void open() throws CatalogException {
+		// test connection, fail early if we cannot connect to database
+		try (Connection conn = DriverManager.getConnection(defaultUrl, username, pwd)) {
+		} catch (SQLException e) {
+			throw new ValidationException(
+				String.format("Failed connecting to %s via JDBC.", defaultUrl), e);
+		}
+
+		LOG.info("Catalog {} established connection to {}", getName(), defaultUrl);
+	}
+
+	@Override
+	public void close() throws CatalogException {
+		LOG.info("Catalog {} closing", getName());
+	}
+
+	// ------ table factory ------
+
+	public Optional<TableFactory> getTableFactory() {
```

Comment:
Do we need overwrite this method, I thought this is a hacky way for hive.

Reply:
no, it's not a hack. This is the standard way to integrate table factory with catalog

Reply:
Why we need this method? Just use SPI is not OK?

Reply:
since it's deterministic that jdbc catalog will work with jdbc table factory, thus planner won't need to go thru service discovery. This is a short cut. When all external systems have their own catalogs, SPI will no long be necessary for table factories.